jax/numpy/lax_numpy.py

```diff
@@ -126,29 +126,40 @@ def __init__(shape, dtype=None, buffer=None, offset=0, strides=None,
 size = onp.size
 _dtype = dtypes.result_type
 
-bool_ = onp.bool_
-int_ = dtypes.int_
-float_ = dtypes.float_
-complex_ = dtypes.complex_
-
-uint8 = onp.uint8
-uint16 = onp.uint16
-uint32 = onp.uint32
-uint64 = onp.uint64
-int8 = onp.int8
-int16 = onp.int16
-int32 = onp.int32
-int64 = onp.int64
-bfloat16 = dtypes.bfloat16
-float16 = onp.float16
-float32 = single = onp.float32
-float64 = double = onp.float64
-complex64 = csingle = onp.complex64
-complex128 = cdouble = onp.complex128
+# At present JAX doesn't have a reason to distinguish between scalars and arrays
+# in its object system. Further, we want JAX scalars to have the same type
+# promotion behaviors as JAX arrays. Rather than introducing a new type of JAX
+# scalar object with JAX promotion behaviors, instead we make the JAX scalar
+# types return JAX arrays when instantiated.
+
+class _ScalarType(object):
+  def __new__(cls, x):
+    return array(x, dtype=cls.dtype)
+
+def _make_scalar_type(onp_scalar_type):
+  return type(onp_scalar_type.__name__, (_ScalarType,),
+              {"dtype": onp.dtype(onp_scalar_type)})
+
+bool_ = _make_scalar_type(onp.bool_)
+uint8 = _make_scalar_type(onp.uint8)
+uint16 = _make_scalar_type(onp.uint16)
+uint32 = _make_scalar_type(onp.uint32)
+uint64 = _make_scalar_type(onp.uint64)
+int8 = _make_scalar_type(onp.int8)
+int16 = _make_scalar_type(onp.int16)
+int32 = _make_scalar_type(onp.int32)
+int64 = _make_scalar_type(onp.int64)
+bfloat16 = _make_scalar_type(dtypes.bfloat16)
+float16 = _make_scalar_type(onp.float16)
```

Comment:
Could you add basic tests for these, e.g., to verify that `x.astype(jax.numpy.float32)` and  `x.astype(jax.numpy.bfloat16)` work where `x` is a JAX or NumPy array, and to verify that you get 0d arrays from direct use? I suppose there might be some existing test coverage for casting, but probably not for 0d arrays?

Reply:
Done. The tests also caught a preexisting bug where `issubdtype(bfloat16, bfloat16)` was false.